replay2/memory_remapper/memory_remapper.h

```diff
@@ -12,4 +12,41 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-void foo(int var);
+#ifndef REPLAY2_MEMORY_REMAPPER_MEMORY_REMAPPER_H
+#define REPLAY2_MEMORY_REMAPPER_MEMORY_REMAPPER_H
+
+#include "address_range.h"
+#include "capture_address.h"
+#include "memory_observation.h"
+#include "replay_address.h"
+
+#include <assert.h>
+#include <map>
+
+#include "replay2/core_utils/non_copyable.h"
+
+namespace agi {
+namespace replay2 {
+
+class MemoryRemapper : public NonCopyable {
+   public:
+    ReplayAddress AddMapping(const MemoryObservation& observation);
+    void RemoveMapping(const CaptureAddress& captureAddress);
+
+    ReplayAddress RemapCaptureAddress(const CaptureAddress& captureAddress) const;
+
+    class AddressNotMappedException : public std::exception {};
+    class AddressAlreadyMappedException : public std::exception {};
+    class RemoveMappingOffsetAddressException : public std::exception {};
+
+   private:
+    std::map<CaptureAddressRange, ReplayAddressRange>::const_iterator CaptureAddressRangeIter(
+        const CaptureAddress& captureAddress) const;
+
+    std::map<CaptureAddressRange, ReplayAddressRange> captureAddressRanges_;
```

Comment:
I have a concern about using an ordered map here.

I know we have talked about hash functions and they are not being free. But they are linear in access time and most of the operations will be access.

Here we have a map that has O(logn) access and there is some iterator search that we use everytime to access. And as far as I can see it uses upper_bound which is another O(logn) function. 

As we don't require any order, I don't know why we are paying the cost of higher algorithmic complexity.

Reply:
We have to use an ordered map here. This map contains a sparse representation of the currently mapped memory. That is, a 1MB mapped region has a single entry in the map, not one entry per byte. This is why we have that call to upper_bound. If we try to remap something 0.7mb of the way though that 1 megabyte region R1: upper_bound returns us the region R2 (or .end()) which comes after R1, we can then wind the iterator back one position to get an interator to R1, and compute the offset between the address we're mapping and R1's first byte. This allows us to bounds check if the address we're remapping is inside R1 or between R1 and R2 in unmapped memory. We can then return the appropriate offset ReplayAddress or throw an error.

Without ordering I don't see any way of maintaining a sub-O(n) time sparse mapping.

Reply:
I see.. This function does completely different work than I expected. Should we have some documentation around this module as it will be commonly used everywhere.

Reply:
Comments added.

Reply:
Comments disappeared 😬